qemu: add usb host passthrough

victortoso opened the pull request:
<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

Hi,

This PR is a WIP related to the request in https://github.com/containers/podman/issues/16707

At this moment, the PR works for `QEMU` hypervisor and allows multiple USB devices to be passthrough to `podman machine`.

Example using `vendor:product`

```
toso@tapioca ~/s/podman (usb-host-passthrough)> ./bin/podman machine init tosovm1 --usb vendor=13d3,product=5406 --usb vendor=08ec,product=0016
toso@tapioca ~/s/podman (usb-host-passthrough)> ./bin/podman machine start tosovm1
toso@tapioca ~/s/podman (usb-host-passthrough)> ./bin/podman machine ssh tosovm1 dmesg | grep -i "13d3\|08ec"
[    1.382442] usb 1-1: New USB device found, idVendor=13d3, idProduct=5406, bcdDevice=60.01
[    1.877370] usb 1-2: New USB device found, idVendor=08ec, idProduct=0016, bcdDevice= 2.00
[   10.948755] usb 1-1: Found UVC 1.10 device Integrated Camera (13d3:5406)
[   10.957727] usb 1-1: Found UVC 1.50 device Integrated Camera (13d3:5406)
t
```

Example using `bus-device_number`

```
toso@tapioca ~/s/podman (usb-host-passthrough)> ./bin/podman machine init tosovm2 --usb bus=1,devnum=3 --usb bus=1,devnum=13
toso@tapioca ~/s/podman (usb-host-passthrough)> ./bin/podman machine start tosovm2
toso@tapioca ~/s/podman (usb-host-passthrough)> ./bin/podman machine ssh tosovm2 dmesg | grep -i "13d3\|08ec"
[    1.401638] usb 1-1: New USB device found, idVendor=13d3, idProduct=5406, bcdDevice=60.01
[    1.893233] usb 1-2: New USB device found, idVendor=08ec, idProduct=0016, bcdDevice= 2.00
[   10.928367] usb 1-1: Found UVC 1.10 device Integrated Camera (13d3:5406)
[   10.934437] usb 1-1: Found UVC 1.50 device Integrated Camera (13d3:5406)
```

Note that this PR does not handle device permissions from the host. Both devices are under my user group, that is:

```
toso@tapioca ~/s/podman (usb-host-passthrough)> whoami
toso
toso@tapioca ~/s/podman (usb-host-passthrough)> groups
toso wheel docker
toso@tapioca ~/s/podman (usb-host-passthrough)> ls -l /dev/bus/usb/001/003
crw-rw-r--. 1 root toso 189, 2 Oct 30 23:30 /dev/bus/usb/001/003
toso@tapioca ~/s/podman (usb-host-passthrough)> ls -l /dev/bus/usb/001/013
crw-rw----. 1 root toso 189, 12 Oct 30 23:50 /dev/bus/usb/001/013
```

Disclaimer: This is my first contribution to Podman. Please do feel free to point out anything that I might be doing wrong.

#### Does this PR introduce a user-facing change?
Yes, it adds a new -usb command line option to `podman machine init`.

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Add USB host passthrough to podman machine (QEMU)
```

victortoso commented:
@ashley-cui, do you think this is a good way of implementing this feature?
 
/cc @beriberikix (for knowledge)

ashley-cui commented:
Thanks for the contribution!

Ah, this seems to get past the issue I was running into when looking at this, which was the host perms. 

 I'd prefer the syntax like `--usb vendorid=3d3,productid=5406`. This is mostly because this makes the syntax more universal for different platforms, as eventually we may need to support different providers that are not qemu.

victortoso commented:
Thanks for the review!

> I'd prefer the syntax like --usb vendorid=3d3,productid=5406. This is mostly because this makes the syntax more universal for different platforms, as eventually we may need to support different providers that are not qemu.

No problem, I'll change it.

Questions that I have:
- Does podman need to handle device's permissions somehow or is that on user's side?
- Applying USB host passthrough on `podman machine init` is enough for now? Just wondering if some sort of hotplug can/should be consider in the future too (e.g: `podman machine init` does not have USB but we could add at `podman machine start`  without storing the devices to config...)

(Asking also to know podman machine scope!)

ashley-cui commented:
> * Does podman need to handle device's permissions somehow or is that on user's side?

I think we would prefer things to be as easy as possible for the user. I've been doing experiments with USB passthrough for the past week and discovered that even if the user changes the permissions of the usb device on /dev/bus/usb, on sleeps, the USB permissions revert. This might end up meaning that we need to let the user deal with perms, since we can't really constantly monitor the device. 

> * Applying USB host passthrough on `podman machine init` is enough for now? Just wondering if some sort of hotplug can/should be consider in the future too (e.g: `podman machine init` does not have USB but we could add at `podman machine start`  without storing the devices to config...)


I think we should definitely consider hotplugging, but I suppose this isn't an initial requirement.  At the very least, I think --usb should also be plumbed through `podman machine set`. This should help out with broken USBs, as I mentioned above, USBs tend to disconnect on sleeps.

victortoso commented:
> This might end up meaning that we need to let the user deal with perms, since we can't really constantly monitor the device.

One possibility is to interact with udev, but that's a linux only solution. I'm sure Window's would have its own APIs in a way that we can manage permissions as well. I don't see podman using udev as of now so we might want to do it as a follow up.

My question was also related to the expectations wrt permissions. We agree that the least the user has to do, the better.

> I think we should definitely consider hotplugging, but I suppose this isn't an initial requirement.

Ok

> At the very least, I think --usb should also be plumbed through podman machine set. This should help out with broken USBs, as I mentioned above, USBs tend to disconnect on sleeps.

Ok, I'll start working on it then.

beriberikix commented:
RE: windows & hotplugging - I think that's a can of worms :worm: 

A neat tool [usbipd-win](https://github.com/dorssel/usbipd-win) had to essentially implement a udev-like thing in Windows. Instead of the native WinUSB stack they opted to leverage the VirtualBox drivers. More discussion [here](https://github.com/dorssel/usbipd-win/discussions/374).

baude commented:
if this PR is only for qemu, then we should try to fence off the option by provider *or* error out in every other provider.

victortoso commented:
> if this PR is only for qemu, then we should try to fence off the option by provider or error out in every other provider.

Makes sense to me. Any suggestions on how to do that? I plan to use a conditional `provider.VMType() == QemuVirt` for exposing `--usb` option. Problem is that we might need some refactor as `provider` is not yet initialized on `cmd/podman/machine/init.go`). I'll tackle this later Today, but any other suggestions are welcome. Thanks for the review!

ashley-cui commented:
I think in the Init() functions in `pkg/machine/<provider>`, for non qemu providers, you can return an err if usb has been set.

victortoso commented:
- Fixed the `make validate` (CI)
- Error out on other hypervisors when using --usb (baude)
- Add some documentation (Ashley)
- Squashed the commits (Ashley)

I need some more time for the e2e test. I need to mock a USB or run the e2e test in VM, not sure what is easier.
I'll back on this only next Monday.

beriberikix commented:
Emulating a USB on Linux is pretty trivial, depending on how much you need to emulate. In the past I've used USB/IP with a little hackery but quick bit of research led me to this neat tool:https://github.com/jtornosm/USBIP-Virtual-USB-Device (details [here](https://jtornosm.fedorapeople.org/usb_device_emulation/).)

victortoso commented:
> Emulating a USB on Linux is pretty trivial, depending on how much you need to emulate. 

Nothing fancy, just to show in the `dmesg` of guest os.

> In the past I've used USB/IP with a little hackery but quick bit of research led me to this neat tool:

I'm not familiar with podman's ci yet, we need to check if kernel has the right flags and it is fine to modprobe those drivers in the host.

In the KubeVirt feature I was working, we solved this with device emulation in QEMU as the CI was running in a VM, we just added an emulated storage to that VM (see https://github.com/kubevirt/kubevirtci/pull/996)

Again, not 100% sure what the best course is for the tests. Mocking with [umockdev](https://github.com/martinpitt/umockdev) might be best but I'm open to suggestions.

ashley-cui commented on docs/source/markdown/podman-machine-init.1.md.in:
```diff
@@ -104,6 +104,18 @@ means to use the timezone of the machine host.
 
+#### **--usb**=*bus=number,devnum=number* or *vendor=hexadecimal,product=hexadecimal*
+
+Assign a USB device from the host to the VM.
+Only supported for QEMU Machines.
+
+The device needs to be present when the VM starts.
+The device needs to have proper permissions in order to be assign to podman machine.
+
+Note that using bus and device number are simpler but the values can change every boot or when the
+device is unplugged. Using vendor and product might lead to collision in the case of multiple
+devices with the same vendor product value, the first available device is assigned.
```
```suggestion
Assign a USB device from the host to the VM via USB passthrough.
Only supported for QEMU Machines.

The device needs to have proper permissions in order to be passed to the machine. This means the device needs to be under your user group.

Note that using bus and device number are simpler but the values can change every boot 
or when the device is unplugged.

When specifying a USB using vendor and product ID's,  if mire than one device has the 
same vendor and product ID, the first available device is assigned.
```

ashley-cui commented on pkg/machine/qemu/machine.go:
```diff
@@ -402,6 +402,14 @@ func (v *MachineVM) Set(_ string, opts machine.SetOptions) ([]error, error) {
 		}
 	}
 
+	if opts.USBs != nil {
```
In the case of `opts.USBs=""`, ie `podman machine set --usb=""`, will the usb be removed from the machine? If so, let's document this.

victortoso replied:
No, that would be considered invalid input with the code we have now. I'll change to your suggestion plus the documentation.

Another option would be another command line argument that states cleaning/reset/disable the usb configuration, eg: `podman machine set --disable-usb`. But only usb would need it, so I'm not sure it is worth to introduce a new command line option, so I agree with `--usb=""` instead.

victortoso replied:
This is enough to fix it. Thanks for the suggestion.

```diff
--- a/pkg/machine/qemu/config.go
+++ b/pkg/machine/qemu/config.go
@@ -66,6 +66,11 @@ func (v *MachineVM) setNewMachineCMD(qemuBinary string, cmdOpts *setNewMachineCM
 func parseUSBs(usbs []string) ([]machine.USBConfig, error) {
        configs := []machine.USBConfig{}
        for _, str := range usbs {
+               if str == "" {
+                       // Ignore --usb="" as it can be used to reset USBConfigs
+                       continue
+               }
+
                vals := strings.Split(str, ",")
                if len(vals) != 2 {
                        return configs, fmt.Errorf("usb: fail to parse: missing ',': %s", str)
```

ashley-cui commented on docs/source/markdown/podman-machine-set.1.md.in:
```diff
@@ -52,6 +52,18 @@ are no longer visible with the default connection/socket. This is because the ro
 users in the VM are completely separated and do not share any storage. The data however is not
 lost and you can always change this option back or use the other connection to access it.
 
+#### **--usb**=*bus=number,devnum=number* or *vendor=hexadecimal,product=hexadecimal*
```
Same as above.

victortoso replied:
I take this is about `podman machine set--usb=""` cleaning up usb configs. I'll improve it shortly.

victortoso replied:
```diff
--- a/docs/source/markdown/podman-machine-set.1.md.in
+++ b/docs/source/markdown/podman-machine-set.1.md.in
@@ -52,13 +52,14 @@ are no longer visible with the default connection/socket. This is because the ro
 users in the VM are completely separated and do not share any storage. The data however is not
 lost and you can always change this option back or use the other connection to access it.

-#### **--usb**=*bus=number,devnum=number* or *vendor=hexadecimal,product=hexadecimal*
+#### **--usb**=*bus=number,devnum=number* or *vendor=hexadecimal,product=hexadecimal* or *""*

 Assign a USB device from the host to the VM.
 Only supported for QEMU Machines.

 The device needs to be present when the VM starts.
 The device needs to have proper permissions in order to be assign to podman machine.
+Use an empty string to remove all previously set USB devices.

 Note that using bus and device number are simpler but the values can change every boot or when the
 device is unplugged. Using vendor and product might lead to collision in the case of multiple
```

ashley-cui reviewed (commented):
A few comments, thanks!

victortoso commented on pkg/machine/hyperv/config.go:
```diff
@@ -166,6 +166,10 @@ func (v HyperVVirtualization) NewMachine(opts machine.InitOptions) (machine.VM,
 		Memory:   opts.Memory,
 	}
 
+	if len(opts.USBs) > 0 {
```
Ok

victortoso commented on pkg/machine/applehv/config.go:
```diff
@@ -137,6 +137,10 @@ func (v AppleHVVirtualization) NewMachine(opts machine.InitOptions) (machine.VM,
 	// Set creation time
 	m.Created = time.Now()
 
+	if len(opts.USBs) > 0 {
```
Ok

victortoso commented on pkg/machine/wsl/config.go:
```diff
@@ -37,6 +38,10 @@ func (p *WSLVirtualization) NewMachine(opts machine.InitOptions) (machine.VM, er
 	vm.ConfigPath = configPath
 	vm.ImagePath = opts.ImagePath
 
+	if len(opts.USBs) > 0 {
```
Ok

victortoso commented:
I had to do a rebase due conflict with https://github.com/containers/podman/commit/a45ba06d020061632de12d48be80eceb520f8307. @baude's patch made me consider if I should move [USBConfig](https://github.com/victortoso/podman/blob/usb-host-passthrough/pkg/machine/config.go#L276) under `machine/define` or not. Not sure. I kept as is to reduce unrequested changes. Just let me know if I should ;)

Cheers,

ashley-cui commented:
LGTM, thanks for all your work!

@containers/podman-maintainers  PTAL

victortoso commented:
I moved the tests to `pkg/machine/qemu/config_test.go` due conflict with rebase. I think it makes sense as this is qemu-only for now.

ashley-cui commented:
@containers/podman-maintainers PTAL, ready to be reviewed :)

ashley-cui commented:
/approve

rhatdan commented on pkg/machine/qemu/command.go:
```diff
@@ -46,6 +47,24 @@ func (q *QemuCmd) SetNetwork() {
+			*q = append(*q, "-device", dev)
+		} else {
+			dev := fmt.Sprintf("usb-host,vendorid=%d,productid=%d", usb.Vendor, usb.Product)
+			*q = append(*q, "-device", dev)
```
Nit: move this line outside of the if block, and  it is only needed once.

victortoso replied:
Ok, the change:

```diff
--- a/pkg/machine/qemu/command.go
+++ b/pkg/machine/qemu/command.go
@@ -55,13 +55,13 @@ func (q *QemuCmd) SetUSBHostPassthrough(usbs []machine.USBConfig) {
        // Add xhci usb emulation first and then each usb device
        *q = append(*q, "-device", "qemu-xhci")
        for _, usb := range usbs {
+               dev := ""
                if usb.Bus != "" && usb.DevNumber != "" {
-                       dev := fmt.Sprintf("usb-host,hostbus=%s,hostaddr=%s", usb.Bus, usb.DevNumber)
-                       *q = append(*q, "-device", dev)
+                       dev = fmt.Sprintf("usb-host,hostbus=%s,hostaddr=%s", usb.Bus, usb.DevNumber)
                } else {
-                       dev := fmt.Sprintf("usb-host,vendorid=%d,productid=%d", usb.Vendor, usb.Product)
-                       *q = append(*q, "-device", dev)
+                       dev = fmt.Sprintf("usb-host,vendorid=%d,productid=%d", usb.Vendor, usb.Product)
                }
+               *q = append(*q, "-device", dev)
```

victortoso commented:
hm, CI is failing but it seems to be on quay.io side.

rhatdan commented:
/lgtm
Thanks @victortoso

flouthoc commented on pkg/machine/qemu/config.go:
```diff
@@ -59,6 +60,78 @@ func (v *MachineVM) setNewMachineCMD(qemuBinary string, cmdOpts *setNewMachineCM
+			return configs, fmt.Errorf("usb: fail to parse: missing '=': %s", str)
+		}
+
+		option := ""
+		if (left[0] == "bus" && right[0] == "devnum") ||
+			(right[0] == "bus" && left[0] == "devnum") {
+			option = "bus_devnum"
+		}
+		if (left[0] == "vendor" && right[0] == "product") ||
+			(right[0] == "vendor" && left[0] == "product") {
+			option = "vendor_product"
+		}
+
+		switch option {
+		case "bus_devnum":
+			bus, devnumber := left[1], right[1]
+			if right[0] == "bus" {
+				bus, devnumber = devnumber, bus
+			}
+
+			configs = append(configs, machine.USBConfig{
+				Bus:       bus,
+				DevNumber: devnumber,
+			})
+		case "vendor_product":
+			vendorStr, productStr := left[1], right[1]
+			if right[0] == "vendor" {
+				vendorStr, productStr = productStr, vendorStr
+			}
+
+			vendor, err := strconv.ParseInt(vendorStr, 16, 0)
+			if err != nil {
+				return configs, fmt.Errorf("fail to convert vendor of %s: %s", str, err)
```
I think `usb:` adds more context and consistent with other error message.

```suggestion
				return configs, fmt.Errorf("usb: fail to convert vendor of %s: %s", str, err)
```

flouthoc replied:
Can this be simplified

```go
option := left[0] +"_"+left[1] // and switch can be match against "vendor_product" and "product_vendor" both.
```

switch

flouthoc commented on pkg/machine/qemu/config.go:
```diff
@@ -59,6 +60,78 @@ func (v *MachineVM) setNewMachineCMD(qemuBinary string, cmdOpts *setNewMachineCM
+
+			product, err := strconv.ParseInt(productStr, 16, 0)
+			if err != nil {
+				return configs, fmt.Errorf("fail to convert product of %s: %s", str, err)
```
Same 
```suggestion
				return configs, fmt.Errorf("usb: fail to convert product of %s: %s", str, err)
```

flouthoc commented:
Created a PR https://github.com/containers/podman/pull/20678

beriberikix commented:
Hi! Is there a specific reason why USB passthrough is not supported on Windows?
![image](https://github.com/containers/podman/assets/178376/4a169622-8480-40a2-a6c6-d7ed7e51d1d6)

rhatdan commented:
@n1hility WDYT?

victortoso commented:
> Hi! Is there a specific reason why USB passthrough is not supported on Windows?

@beriberikix you should probably open a new issue for WSL

beriberikix commented:
Does Podman on Windows support QEMU or only WSL? USB support on WSL is [complicated](https://blog.golioth.io/program-mcu-from-wsl2-with-usb-support/) , so targeting QEMU would be ideal!

rhatdan commented:
We support hyperV and WSL no QEMU
